Fix #4275 fail URIs with ambiguous segments

gregw opened the pull request:
Fix #4275 fail URIs with ambiguous segments.

gregw committed:
Fix #4275 fail URIs with ambiguous segments

Fix #4275 fail URIs with ambiguous segments.

gregw commented:
@sbordet @joakime nudge

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1815,6 +1817,18 @@ public void setMetaData(org.eclipse.jetty.http.MetaData.Request request)
+            if (!allow)
+                throw new BadMessageException("Ambiguous segment in URI");
+        }
+
```
No tests for this code?

gregw replied:
Test added. I also inverted the sense of the compliance to better align with others.

sbordet reviewed (changes requested):
Since we are changing `HttpURI` parsing, I think we should go the extra mile and do it as per RFC3986.

https://tools.ietf.org/html/rfc3986#section-3.3 says that even the `,` can be used as separator between a segment and a segment-parameter.

But if we go to the ABNF, possibly any `sub-delims` can be present, and we should not consider those as part of the segment.
Doing so would probably require another compliance mode.

I changed `testAmbiguousSegments()` and using a `,` instead of `;` fails the test.

gregw commented:
@sbordet I don't see why we should treat ',' as a parameter separator? It might be used as such in some schemes, but not in HTTP.  We will never split a segment by ',', but we do split by ';'.  I think we need to be absolutely consistent with our handling to avoid issues, so if we start splitting on ',' here, we'd have to split everywhere and I don't think that is valid to do.

We are treating `..;` as ambiguous because the RFC says we should only normalise `..` and not `..;`, yet we know we later split on `;` so that only `..` is left, which an application might normalise or might give back to us and we will normalise. I don't see the same problem with `..,`.

gregw committed:
updates from review

Invert sense of ambiguous path segment compliance
added test

sbordet commented:
> I don't see why we should treat ',' as a parameter separator?

Because the RFC says:

> For example, the semicolon (";") and equals ("=") reserved characters are
   often used to delimit parameters and parameter values applicable to
   that segment.  The comma (",") reserved character is often used for
   similar purposes.  For example, one URI producer might use a segment
   such as "name;v=1.1" to indicate a reference to version 1.1 of
   "name", whereas another might use a segment such as "name,1.1" to
   indicate the same.

So if `;` and `,` are interchangeable, and we only fix `;`, then we'll have the same problem when someone uses `,`, no?
And the same for all the other `sub-delims` characters.

The fact that it's not used _typically_ in HTTP does not mean pentesters can't generate such URIs to validate URI handling.

gregw commented:
`,` and `;` are not interchangeable. 
I known of no HTTP "URI producer" or consumer that will treat `,` in a segment as special.

We only have an issue with HTTP schema URIs as that is all we handle.  We never split on `,`, we do not remove `,param`.    I see no reason to not allow `..,` as a segment.  Sure it is a silly looking segment, but then there are many other silly segments that we don't block.

Remember, we are not disallowing `..;` because the RFC tells us to. We are disallowing it because the RFC tells us that it is a normal path segment and not a relative one, yet we know that we later strip the `;` and thus can treat it as special.   None of that holds for `,`

joakime commented:
There are multiple things to consider here.

First there's **Reserved Characters** - https://tools.ietf.org/html/rfc3986#section-2.2
Which defines the sub-delims as ...

```
      reserved    = gen-delims / sub-delims

      gen-delims  = ":" / "/" / "?" / "#" / "[" / "]" / "@"

      sub-delims  = "!" / "$" / "&" / "'" / "(" / ")"
                  / "*" / "+" / "," / ";" / "="
```

Which means `,` and `;` are definitely equivalent.

The following example URIs should be considered in this PR.

```
/a/..;/b/
/a/..;foo/b/
/a/..;foo=bar/b/
/a/..,bar/b/
/a/..,foo=bar/b/
/a/..$/b/
/a/..!/b/
/a/..&/b/
/a/.."/b/
/a/..()/b/
/a/..*/b/
/a/..+/b/
/a/..=bar/b/
```

Then we have **Remove Dot Segments** - https://tools.ietf.org/html/rfc3986#section-5.2.4
Which is part of the resolution process for URIs.
Which pretty much tells us that `.` and `..` are to be removed, even if they have parameters.

This is getting quite complicated, and I'm wondering if we REALLY need to support URI parameters at all?

I have been unable to find another web server that supports them.
Even Apache httpd doesn't support them.

Two examples:

```
# These work
$ curl https://en.wikipedia.org/wiki/Jetty
$ curl https://httpd.apache.org/docs/2.4/mod/mod_rewrite.html

# These do not work as expected
$ curl https://en.wikipedia.org/wiki;/Jetty
# above results in a redirect to the mainpage
$ curl https://httpd.apache.org/docs/2.4/mod;/mod_rewrite.html
# above results in a 404 status response
```

joakime commented:
An interesting read on a complex / contrived URI example.

https://techdocs.broadcom.com/us/en/ca-enterprise-software/devops/devtest-solutions/10-4/using/using-ca-service-virtualization/ca-service-virtualization-concepts/parsing-uris-in-http-requests.html

joakime commented:
> This is getting quite complicated, and I'm wondering if we REALLY need to support URI parameters at all?

I retract this statement, I've found many examples in REST libraries that use various forms of the reserved characters in the URI to mean special things to the REST library.

Some even making the "contrived" example in the devtest link look easy to follow.

gregw commented:
@sbordet and @joakime, you are over complicating this and missing the actual issue!

Our URI handling is correct according to both the RFC and the servlet spec.    However, there is a problem between those two spec.   The RFC says that `;` might be a special character, yet does not say it should be stripped at the URI level.   However the servlet spec expects that path parameters will be stripped.        Our code is aware of this and we treat mapping of both security constraints and servlets in exactly the same way, so there can be no bypasses.     However, because we strip parameters from the URI before giving them to the servlet, then we throw away information and the application can get a path with a `..` segment.   It should know that this is not really a `..` segment and that it came either from `%2e%2e` or `..;param`.   However, it is pretty easy for the application to forget this an normalise it, or not re-encode it and give it back to us, so we normalise it.

This is ONLY a problem with `;` and % encoding, because they are the only two ways that we can ever decode a URI and give the application a `..` segment.    The solution I have coded here is to just not support `%2e%2e` and `..;param` (plus a few variations of those).    If there are some other URIs that we also decode to result in `..` segments, then we need to handle those as well, but none of the examples either of you have listed appear to be such examples.

gregw committed:
Merge remote-tracking branch 'origin/jetty-9.4.x' into jetty-9.4.x-4275-ambiguous-path-segments

gregw commented:
Note also that the code doesn't actually treat `;` that specially.  It has modified the URI parser  so that any character than ends a segment (`;`, `?`, `'#`, etc.) will result in the segment being check as potentially ambiguous.  

However, I will double check how we handle things like `/foo/..?bar` ...

gregw commented:
OK, I think I have found one more related issue.  The RFC says the URI is:
```abnf
URI         = scheme ":" hier-part [ "?" query ] [ "#" fragment ]
```
We correctly normalise `/foo/bar/..?query` to `/foo#query`, but we don't do  `/foo/bar/..#fragment` to `/foo#fragment`. Thus when we strip the fragment later, we may get a similar problem.
So I think our URIUtil canonical path methods should probably also handle `#` just like `?`, even though that should never come over the wire.

gregw committed:
updates from review

Fixed tests for `%2e%2e`

gregw committed:
updates from review of RFC

 + Correctly handle canonical trailing '.' and '..'
 + Perform canonical path resolution before decoding
 + Treat decoding of %2f as an ambiguous segment.
 + better tests of ambiguous segments

joakime commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -65,6 +67,30 @@
+    static
+    {
+        __ambiguousSegments.put("%2e", Boolean.TRUE);
+        __ambiguousSegments.put("%2e%2e", Boolean.TRUE);
```
This specific case `%2e%2e` I have especially difficult time accepting.
As that can show up as valid.
We even have testcases for it in DefaultServletTest. (see "testListingContextBreakout")
When the segment is 100% encoded I think it's fair and legit use.
However I do agree that when it's mixed (encoded / not-encoded) it's not valid and ambiguous.

gregw replied:
They are all fair usages.  They are just ambiguous because when we give a servlet a string version, we have thrown away information.
I think it is reasonable for a webapp that wants to consume `%2e%2e` to turn on the compliance mode to say they have their big boy pants on and are prepared to deal with that ambiguity.  But if an app doesn't do that, then it is best to not allow that request through.

gregw replied:
Too be more precise, I don't think we should be making a distinction between very ambiguous and slightly ambiguous URIs.   The compliance mode should simply exclude URIs that once we decode them and strip parameters it is impossible to recover exactly what the canonical segment boundaries are.

For me, the switching off that mode should be done by applications that will look at the raw getRequestURI and do their own parsing and decoding.  Note also with this change, we no longer strip intermediary parameters from the raw getRequestURI, so apps that want those can now get them.

joakime reviewed (commented):
I like the majority of this PR, I just don't like one aspect of the ambiguousSegments list.

gregw committed:
updates from review of RFC

 + Fixed test failures that were using %2f

gregw committed:
Fix test

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@sbordet nudge

gregw commented on jetty-http/src/test/java/org/eclipse/jetty/http/HttpURITest.java:
```diff
@@ -221,4 +227,124 @@ public void testBasicAuthCredentials() throws Exception
         assertEquals(uri.getUser(), "user:password");
     }
+
+    public static Stream<String> pathsWithAmbiguousSegments()
```
I think I will combine all these tests and add a few more comments

gregw committed:
updates from review of RFC

 + simplified tests

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java:
```diff
@@ -62,7 +62,7 @@
      * {@link HttpComplianceSection#MULTIPLE_CONTENT_LENGTHS},
      */
-    RFC2616_LEGACY(sectionsBySpec("RFC2616,-FIELD_COLON,-METHOD_CASE_SENSITIVE,-TRANSFER_ENCODING_WITH_CONTENT_LENGTH,-MULTIPLE_CONTENT_LENGTHS")),
+    RFC2616_LEGACY(sectionsBySpec("RFC2616,-FIELD_COLON,-METHOD_CASE_SENSITIVE,-TRANSFER_ENCODING_WITH_CONTENT_LENGTH,-MULTIPLE_CONTENT_LENGTHS,-NO_AMBIGUOUS_PATH_SEGMENTS")),
```
Update javadocs.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java:
```diff
@@ -72,7 +72,7 @@
      * Jetty's current RFC7230 support, which incorrectly excludes  {@link HttpComplianceSection#METHOD_CASE_SENSITIVE}
      */
-    RFC7230_LEGACY(sectionsBySpec("RFC7230,-METHOD_CASE_SENSITIVE")),
+    RFC7230_LEGACY(sectionsBySpec("RFC7230,-METHOD_CASE_SENSITIVE,-NO_AMBIGUOUS_PATH_SEGMENTS")),
```
Update javadocs.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -77,6 +103,8 @@
     String _uri;
     String _decodedPath;
 
+    boolean _ambiguousSegment;
```
Please review copy-constructor `HttpURI(HttpURI uri)` as it does not copy this new field, nor `_user`, nor `_decodedPath`.

sbordet replied:
Make the above 3 fields private and compact them with the others.

sbordet replied:
Method `clear()` does not null-out field `_user`. Perhaps move `clear()` close to the fields?

sbordet replied:
This class implements `equals()` but not `hashCode()`! Must be implemented using the same fields as `equals()`.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -486,45 +542,37 @@ else if (c == '/')
             case QUERY:
                 _query = uri.substring(mark, end);
                 break;
         }
```
Please add `default` with explicit `break` if that's the case, or throwing if it should not be reached.

sbordet commented on jetty-server/src/test/java/org/eclipse/jetty/server/HttpConnectionTest.java:
```diff
@@ -500,43 +500,42 @@ public void testOKPathDotDotPath() throws Exception
         String response = connector.getResponse("GET /ooops/%2e%2e/path HTTP/1.0\r\nHost: localhost:80\r\n\n");
-        checkContains(response, 0, "HTTP/1.1 200 OK");
-        checkContains(response, 0, "pathInfo=/path");
+        checkContains(response, 0, "HTTP/1.1 400 ");
```
Change the method name, this test is now "Bad".

So this is actually a behavioral change that may break people out there. Is it unavoidable? Perhaps use a default in `HttpCompliance` that permits it?

gregw replied:
It is a change in behaviour, but one that we want to do.  Users can get the old behaviour with the compliance mode.
I'm removing these tests from here, since they no longer test what they were, as a %2e%2e can't navigate at all, let alone above the URI root.   The encoding itself is well tested elsewhere.

sbordet commented on jetty-server/src/test/java/org/eclipse/jetty/server/RequestTest.java:
```diff
@@ -1836,6 +1836,31 @@ public void testGetterSafeFromNullPointerException()
+        assertThat(_connector.getResponse(request), startsWith("HTTP/1.1 200"));
+
+
+
```
Remove empty newlines!

sbordet commented on jetty-servlet/src/test/java/org/eclipse/jetty/servlet/AsyncContextTest.java:
```diff
@@ -229,7 +229,7 @@ public void testDispatchAsyncContext() throws Exception
     public void testDispatchAsyncContextEncodedUrl() throws Exception
     {
-        String request = "GET /ctx/test/hello%2fthere?dispatch=true HTTP/1.1\r\n" +
+        String request = "GET /ctx/test/hello%20there?dispatch=true HTTP/1.1\r\n" +
```
Why was this test changed from `%2F` to `%20`? Seems like a big difference not relevant to this PR?

gregw replied:
The test is that we can dispatch to encoded paths correctly. We now treat %2f as an ambiguous path, so using %20 instead gives us the same test but without needing to switch compliance mode.

sbordet commented on jetty-servlet/src/test/java/org/eclipse/jetty/servlet/AsyncContextTest.java:
```diff
@@ -253,16 +253,16 @@ public void testDispatchAsyncContextEncodedUrl() throws Exception
     public void testDispatchAsyncContextSelfEncodedUrl() throws Exception
     {
-        String request = "GET /ctx/self/hello%2fthere?dispatch=true HTTP/1.1\r\n" +
+        String request = "GET /ctx/self/hello%20there?dispatch=true HTTP/1.1\r\n" +
```
Ditto.

gregw replied:
ditto

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -561,10 +635,19 @@ public String getPath()
+            if (canonical == null)
+                throw new BadMessageException("Bad URI");
+            _decodedPath = URIUtil.decodePath(canonical);
+        }
```
I don't like throwing when calling this method.
If the `HttpURI` is invalid, we should fail at `parse()` and not even create the `HttpURI` object, otherwise it seems a good object until I happen to call `getDecodedPath()` where it throws because it was bad.

gregw replied:
It is done here to be lazy in case the decoded path is never needed, but I guess it is fine to spend a little more effort on the minority of URIs that actually need to be decoded.

gregw replied:
@sbordet this has kind of resulted in a bigger refactor of this class than just fixing the problem... but I think it is more consistent now, so like it better.... we just need to really carefully review.
parse now always sets the decodedPath. The other big change is that most of the constructors now call parse, even if just to parse the path. Thus we are checking for ambiguous paths and bad canonical paths in more situations.

gregw replied:
@sbordet Note also that we have refactored this class in 10 to be immutable, so if we fix too much more here the merge is going to be difficult.

gregw committed:
Updates from review

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Updates from review

Signed-off-by: Greg Wilkins <gregw@webtide.com>

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -108,16 +134,25 @@ public HttpURI(String scheme, String host, int port, String path, String param,
         _host = host;
         _port = port;
-        _path = path;
+        parse(State.PATH, path, 0, path.length());
```
`path` may be null, so we need a null guard or we NPE -- I think passing `null` as path should be accepted.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -486,51 +561,84 @@ else if (c == '/')
+        }
+    }
+
+    private void decodePath()
```
This needs to be removed.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1826,7 +1842,7 @@ public void setMetaData(org.eclipse.jetty.http.MetaData.Request request)
         else if (encoded.startsWith("/"))
         {
-            path = (encoded.length() == 1) ? "/" : URIUtil.canonicalPath(uri.getDecodedPath());
+            path = (encoded.length() == 1) ? "/" : uri.getDecodedPath();
```
Few lines below:

```java
if (path == null || path.isEmpty())
{
    setPathInfo(encoded == null ? "" : encoded);
    throw new BadMessageException(400, "Bad URI");
}
```

Seems to me that `path` could be legitimately null at that point, and if not, why do we set the pathInfo, just to throw the next line?

gregw replied:
I don't think we can continue with a null path.  However, we still set the path info before throwing so we don't break any error page handling and/or request logging that might get confused by a null or empty request path.

gregw replied:
Also remember this is cleaned up a lot in 10, so I don't want to fix too many unrelated issues here.

sbordet replied:
Okay then.

gregw committed:
updates from review

updates from review

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -685,10 +685,14 @@ public String getParam()
     {
-        _param = param;
-        if (_path != null && !_path.contains(_param))
+        if (!Objects.equals(_param,param))
```
Space after comma.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -134,10 +135,14 @@ public HttpURI(String scheme, String host, int port, String path, String param,
+        if (query != null)
+            _query = query;
+        if (fragment != null)
+            _fragment = fragment;
```
I would not do null checks for just assignments.
Path needs null check because `path.length()` may NPE.

gregw replied:
I'm doing null checks because if we parse the path, we may discover params, fragments and queries in it.    So either we through if they are non null, or just not overwrite them unless we have explicit non null values.

gregw committed:
updates from review

updates from review
